design/secrets_provider_milestone1_design.md

```diff
@@ -0,0 +1,385 @@
+
+```pseudocode
+for each k8s-secret
+  variable-ids = k8s-secret.parseConjurMap() # list of paths of key in ConjurMap
+  GET /secrets{?variable-ids}
+  # Log failure or success
+```
+
+Variable IDs will resemble the following:
+
+```
+apiVersion: v1
+kind: Secret
+metadata:
+  name: db-credentials
+type: Opaque
+stringData:
+  conjur-map: |-
+    username: secrets/db_username # the value is one element in the variable-ids list
+    password: secrets/db_password # the value is one element in the variable-ids list
+```
+
+#### Fetching Relevant K8s Secrets
+
+With the decoupling of Secrets Provider and application pod, we can know what types of K8s Secrets the pod requires by either of the following ways: 
+
+|      | Fetch/Update K8s Secret                                      | Elaboration                                                  | Pros                                                         | Cons                                        |
+| ---- | ------------------------------------------------------------ | ------------------------------------------------------------ | ------------------------------------------------------------ | ------------------------------------------- |
+| 1    | Read all K8s Secrets that are marked with labels and update them with Conjur secrets |                                                              | - Scalable <br /><br />- Less intervention on our part as a customer can define which types of secrets we should parse<br /><br />- K8s native, as-is solution | - Not generalized so we are still K8s bound |
+| 2    | Get K8s secrets name by configuration in Secrets Provider    | Define K8S_SECRETS in `cyberark-secrets-provider` manifest<br />` name: K8S_SECRET` | - No code changes<br /><br />- Backwards compatible          | - Repetitive work<br /><br />- Not scalable |
+
+To know which K8s Secrets the apps in the namespace need we chose the first solution, Read all K8s Secrets that are marked with labels. It would be the most K8s native and avoids the repetitive work of having to list all the K8s Secrets each app needs.
+
+It is important to note however, that we are responsible for providing a solution for updating the K8s secrets and *not* how those Pods will get those secrets.
+
+##### **Customer Experience**
+
+K8s offers the option of adding labels on K8s secrets. For example, a customer would add a label on a K8s Secret like so:
+
+```yaml
+apiVersion: v1
+kind: Secret
+metadata:
+  name: db-secret
+  labels:
+    configurable-key: env-prod # label name can be configurable
```

Comment:
what if instead we add 
```
labels:
    conjur-secret: true
```

and filter the secrets with the `conjur-secret` label? this way we don't need to maintain the `K8S_SECRET_LABEL` env var in the secrets-provider.

Do we gain anything by having the label name configurable?

Reply:
the idea behind the label was that a customer could already be using labels and so they don't have to add more conjur meta-data to their k8s secret. We give the customer the freedom to use their own labels

Reply:
do we know this is the case? can we maybe have a default value from `K8S_SECRET_LABEL` so customers don't have to add an environment variable?